internal/controller/engine/cache.go

```diff
@@ -0,0 +1,195 @@
+
+	mx     sync.RWMutex
+	active map[schema.GroupVersionKind]bool
+}
+
+// TrackInformers wraps the supplied cache, adding a method to query which
+// informers are active.
+func TrackInformers(c cache.Cache, s *runtime.Scheme) *InformerTrackingCache {
+	return &InformerTrackingCache{
+		Cache:  c,
+		scheme: s,
+		active: make(map[schema.GroupVersionKind]bool),
+	}
+}
+
+// ActiveInformers returns the GVKs of the informers believed to currently be
+// active. The InformerTrackingCache considers an informer to become active when
+// a caller calls Get, List, or one of the GetInformer methods. It considers an
+// informer to become inactive when a caller calls the RemoveInformer method.
+func (c *InformerTrackingCache) ActiveInformers() []schema.GroupVersionKind {
+	c.mx.RLock()
+	defer c.mx.RUnlock()
+
+	out := make([]schema.GroupVersionKind, 0, len(c.active))
+	for gvk := range c.active {
+		out = append(out, gvk)
+	}
+	return out
+}
+
+// Get retrieves an obj for the given object key from the Kubernetes Cluster.
+// obj must be a struct pointer so that obj can be updated with the response
+// returned by the Server.
+//
+// Getting an object marks the informer for the object's GVK active.
+func (c *InformerTrackingCache) Get(ctx context.Context, key client.ObjectKey, obj client.Object, opts ...client.GetOption) error {
+	if err := c.markActive(obj); err != nil {
+		return err
+	}
+
+	return c.Cache.Get(ctx, key, obj, opts...)
+}
+
+// List retrieves list of objects for a given namespace and list options. On a
+// successful call, Items field in the list will be populated with the result
+// returned from the server.
+//
+// Listing objects marks the informer for the object's GVK active.
+func (c *InformerTrackingCache) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	if err := c.markActive(list); err != nil {
+		return err
+	}
+
+	return c.Cache.List(ctx, list, opts...)
+}
+
+// GetInformer fetches or constructs an informer for the given object that
+// corresponds to a single API kind and resource.
+//
+// Getting an informer for an object marks the informer as active.
+func (c *InformerTrackingCache) GetInformer(ctx context.Context, obj client.Object, opts ...cache.InformerGetOption) (cache.Informer, error) {
+	if err := c.markActive(obj); err != nil {
+		return nil, err
+	}
+	return c.Cache.GetInformer(ctx, obj, opts...)
+}
+
+// GetInformerForKind is similar to GetInformer, except that it takes a
+// group-version-kind, instead of the underlying object.
+//
+// Getting an informer marks the informer as active.
+func (c *InformerTrackingCache) GetInformerForKind(ctx context.Context, gvk schema.GroupVersionKind, opts ...cache.InformerGetOption) (cache.Informer, error) {
+	obj := &unstructured.Unstructured{}
+	obj.SetAPIVersion(gvk.GroupVersion().String())
+	obj.SetKind(gvk.Kind)
+
+	if err := c.markActive(obj); err != nil {
+		return nil, err
+	}
+	return c.Cache.GetInformerForKind(ctx, gvk, opts...)
```

Comment:
this feels racy:
- go routine A: markActive
- go routine B: markInactive
- go routine B: RemoveInformer
- go routine A: GetInformerForKind

Reply:
Good catch, will update.

Reply:
Pushed an update that should address this.